File: week6/05-sentimental-credit/credit.py
from cs50 import get_float, get_int, get_string
from math import pow


def main():
    credit()


def digitsToList(number: int) -> list:
    return [int(digit) for digit in str(number)]


def luhn(number: int) -> bool:
    numberlist = digitsToList(number)
    odd = sum([int(num) for num in numberlist[-1::-2]])
    even = sum([sum(digitsToList(int(num) * 2)) for num in numberlist[-2::-2]])
    return (odd + even) % 10


def credit():
    number = get_int("Number: ")
    numberlen = len(str(number))
    if not (luhn(number)):
        if str(number)[0] in ["4"] and (numberlen == 13 or numberlen == 16):
            print("VISA")
        elif str(number)[0:2] in ["34", "37"] and numberlen == 15:
            print("AMEX")
        elif str(number)[0:2] in ["51", "52", "53", "54", "55"] and (
            numberlen == 15 or numberlen == 16
        ):
            print("MASTERCARD")
        else:
            print("INVALID")
    else:
        print("INVALID")


main()

File: week1/05-credit/credit.c
#include <cs50.h>
#include <malloc.h>
#include <stdio.h>
#include <string.h>

// Prototype
int longlen(long n);
bool startMatch(long card_num, int start);
int *longToArray(long n);
void reverseArray(int arr[], int start, int end);
int luhn(long card);

int main(void)
{
    long card = get_long("Number: ");

    // Check card validity
    if (longlen(card) == 15 && (startMatch(card, 34) || startMatch(card, 37)) && luhn(card))
    {
        printf("AMEX\n");
    }
    else if ((longlen(card) == 16) &&
             (startMatch(card, 51) || startMatch(card, 52) || startMatch(card, 53) || startMatch(card, 54) ||
              startMatch(card, 55)) &&
             luhn(card))
    {
        printf("MASTERCARD\n");
    }
    else if ((longlen(card) == 13 || longlen(card) == 16) && startMatch(card, 4) && luhn(card))
    {
        printf("VISA\n");
    }

    else
    {
        printf("INVALID\n");
    }
}

// Length of long
int longlen(long n)
{
    int length = 0;
    while (n > 0)
    {
        n = n / 10;
        length++;
    }

    return length;
}

// Match starting number(s)
bool startMatch(long card, int start)
{
    while (longlen(card) > longlen(start))
    {
        card = card / 10;
    }

    return (card == start);
}

// Creates an array from long in reverse
int *longToArray(long number)
{
    int *arr = malloc(sizeof(long) * longlen(number));
    int i = 0;
    while (number != 0)
    {
        arr[i] = number % 10;
        number /= 10;
        i++;
    }

    return arr;
}

// Gets the sum of the digits of a number
int sumOfDigits(int number)
{
    int sum = 0;
    while (number != 0)
    {
        sum += number % 10;
        number /= 10;
    }

    return sum;
}

// Luhn's Algorithm
int luhn(long card)
{
    // Create a reverse array of the card
    int *cardArr = longToArray(card);

    // Sum of the (sum of the digits of twice the even idx) and odd idx
    int sum = 0;
    for (int i = 0; i < longlen(card); i++)
    {
        if (i & 1)
        {
            sum += sumOfDigits(cardArr[i] * 2);
        }
        else
        {
            sum += cardArr[i];
        }
    }

    // Last digit is 0
    return (sum % 10 == 0);
}
